presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java

```diff
@@ -205,8 +216,24 @@ public Response retryFailedQuery(
                 dispatchManager,
                 queryResultsProvider,
                 retryCount);
+
+        // optimistically register the new query
         queries.put(query.getQueryId(), query);
 
+        // hook the new query to the old query id to uniquely identify the new retry query
+        Query existingRetryQuery = retriedQueries.putIfAbsent(queryId, query);
+        if (existingRetryQuery != null) {
+            // other thread has already create the new retry query
+            // remove the registered new query to reuse the existing one
+            queries.remove(query.getQueryId());
+            query = existingRetryQuery;
+        }
```

Comment:
If there's two or more server-side retries (meaning, sever has requested client to send a retry), how will this look different from client-side retries (meaning, the client retries this GET)?

I also am not sure of the intent of removing the query from the `queries` map--I believe this will cause a leak.

My guess what should happen is this: we can include the retryCount in the path.  We can remove the `retriedQueries` map and just `compute` the new value for the retry.  If, inside the callback for `compute`, the old, existing query has a retryCount that is equal or greater to the retryCount in the path, then do nothing, otherwise return the new query (which will cause the `queries` map to be updated).

Reply:
I double checked the logic. Actually it's more complicated. For any query, the "retry count" will always be one: suppose Q1 failed, then we will send GET to /retry/Q1 and generates Q2. If further Q2 fails, we will send GET to /retry/Q2 and generates Q3, and so on. I started this sequence so we can have the chance to create Q_{n+1} when Q_{n} has not yet been purged. If we go with Q1 with retry count, then we will need other complicated bookkeeping data structure to make sure Q1 is not purged before we exhaust all chances. Note that to difference if we have exhausted all chances is fairly complicated in `Query::getNextResultWithRetry`.

 So I made a bit different change with synchronized method still keeping the retryQueries. Let me know if that looks ok.